src/xml_parser_main_metadata_together.c

```diff
@@ -0,0 +1,371 @@
+        return code;
+    }
+    other_f = cr_open(other_path, CR_CW_MODE_READ, CR_CW_AUTO_DETECT_COMPRESSION, &tmp_err);
+    if (tmp_err) {
+        int code = tmp_err->code;
+        g_propagate_prefixed_error(err, tmp_err, "Cannot open %s: ", other_path);
+        return code;
+    }
+
+    cr_ParserData *primary_pd;
+    //TODO(amatej): replace do_files 0 by check if we are parsing filelists..
+    //TODO(amatej): if newpkgcb return NULL we should skip the pkg...
+    primary_pd = new_primary_parser_data(newpkgcb_primary, &cbdata, primary_pkgcb, &cbdata, warningcb, warningcb_data, 0);
+    cr_ParserData *filelists_pd;
+    filelists_pd = new_filelists_parser_data(newpkgcb_general, &cbdata, filelists_pkgcb, &cbdata, warningcb, warningcb_data);
+    cr_ParserData *other_pd;
+    other_pd = new_other_parser_data(newpkgcb_general, &cbdata, other_pkgcb, &cbdata, warningcb, warningcb_data);
+
+    char primary_is_done = 0;
+    char filelists_is_done = 0;
+    char other_is_done = 0;
+    cbdata.primary_count = 0;
+    cbdata.filelists_count = 0;
+    cbdata.other_count = 0;
+    while (1) {
+        if (!primary_is_done) {
+            do {
+                primary_is_done = parse_next_section(primary_f, primary_path, primary_pd, err);
+                if (*err) {
+                    ret = (*err)->code;
+                    goto out;
+                }
+
+            //TODO(amatej): what about not providing filelists? well at least one of them should always be provided
+            } while (cbdata.primary_count < cbdata.filelists_count || cbdata.primary_count < cbdata.other_count);
+        }
+
+        if (!filelists_is_done) {
+            do {
+                filelists_is_done = parse_next_section(filelists_f, filelists_path, filelists_pd, err);
+                if (*err) {
+                    ret = (*err)->code;
+                    goto out;
+                }
+            } while (cbdata.filelists_count < cbdata.primary_count || cbdata.filelists_count < cbdata.other_count);
+        }
+
+        if (!other_is_done) {
+            do {
+                other_is_done = parse_next_section(other_f, other_path, other_pd, err);
+                if (*err) {
+                    ret = (*err)->code;
+                    goto out;
+                }
+            } while (cbdata.other_count < cbdata.filelists_count || cbdata.other_count < cbdata.primary_count);
+        }
+
+        if (primary_pd->err) {
+            ret = primary_pd->err->code;
+            g_propagate_error(err, primary_pd->err);
+            break;
+        }
+        if (filelists_pd->err) {
+            ret = filelists_pd->err->code;
+            g_propagate_error(err, filelists_pd->err);
+            break;
+        }
+        if (other_pd->err) {
+            ret = other_pd->err->code;
+            g_propagate_error(err, other_pd->err);
+            break;
+        }
+
+        //TODO(amatej): add tests for out of order metadata
```

Comment:
For what it's worth, as mentioned in the email we're not aware of any tools that can generate such metadata, and Pulp 2 would not have been capable of parsing it at all.  We've not had any reports of that issue, so such out-of-order metadata may not exist in the wild.